File: server/controllers/content-controller.js
const asyncHandler = require('express-async-handler');
const Content = require('../models/contentModel');

// @desc    Save or update notepad and whiteboard content
// @route   POST /api/content/save
// @access  Private
const saveContent = asyncHandler(async (req, res) => {
  const { videoUrl, notepadText, canvasImage } = req.body;
  const userId = req.user._id;

  const content = await Content.findOneAndUpdate(
    { videoUrl, user: userId },
    { notepadText, canvasImage },
    { upsert: true, new: true }
  );

  if (!videoUrl || !videoUrl.trim()) {
  res.status(400);
  throw new Error('Video URL is required');
}
  res.json(content);
});

// @desc    Get content for a video URL (by owner or another user)
// @route   GET /api/content
// @access  Private
const getContent = asyncHandler(async (req, res) => {
  const { videoUrl, userId } = req.query;
  const targetUser = userId || req.user._id;

  const content = await Content.findOne({
    videoUrl,
    user: targetUser,
  });

  if (!content) {
    res.status(404);
    throw new Error('No content found for this URL and user.');
  }

  res.json(content);
});

module.exports = { saveContent, getContent };

File: server/app/controllers/content_controller.py
from bson import ObjectId

from fastapi import HTTPException

from pymongo import ReturnDocument

from app.core.database import content_collection
from app.schemas.content import ContentSaveRequest

async def save_content(payload: ContentSaveRequest, current_user):
    
    if not payload.videoUrl.strip():
        raise HTTPException(
            status_code=400,
            detail="Video URL is required"
        )
    content = await content_collection.find_one_and_update(
        {
            "videoUrl": payload.videoUrl,
            "user": current_user["_id"]
        },
        {
          "$set":{
              "notepadText": payload.notepadText,
              "canvasImage": payload.canvasImage
          }
        },
        upsert=True,
        return_document=ReturnDocument.AFTER
    )
    
    content["_id"] = str(content["_id"])
    content["user"] = str(content["user"])
    
    return content

async def get_content(videoUrl: str, current_user, userId: str | None = None):
    target_user = (
        ObjectId(userId)
        if userId
        else current_user["_id"]
    )
    
    content = await content_collection.find_one({
        "videoUrl": videoUrl,
        "user": target_user
    })
    
    if not content:
        raise HTTPException(
            status_code=404,
            detail="No content found for this url by the user"
        )
        
    content["_id"] = str(content["_id"])
    content["user"] = str(content["user"])
    
    return content

async def get_shared_content(videoUrl: str, userId: str):
    content = await content_collection.find_one({
        "videoUrl": videoUrl,
        "user": ObjectId(userId)
    })
    
    if not content:
        raise HTTPExpection(
            status_code=404,
            detail="No content found for this URL and user"
        )
        
    content["_id"] = str(content["_id"])
    content["user"] = str(content["user"])
    
    return content

